core/embed/rust/src/ui/shape/jpeg.rs

```diff
@@ -0,0 +1,196 @@
+                    // Return true if we are not done yet
+                    mcu_r.x1 < clip.x1 || mcu_r.y1 < clip.y1
+                }
+            ),
+            "Invalid JPEG"
+        );
+    }*/
+
+    // This is a little bit slower implementation suitable for ProgressiveRenderer
+    fn draw(&mut self, canvas: &mut dyn Canvas, cache: &DrawingCache<'a>) {
+        let bounds = self.bounds(cache);
+        let clip = canvas.viewport().relative_clip(bounds).clip;
+
+        // Translate clip to JPEG relative coordinates
+        let clip = clip.translate(-canvas.viewport().origin);
+        let clip = clip.translate((-bounds.top_left()).into());
+
+        if self.blur_radius == 0 {
+            // Draw JPEG without blurring
+
+            // Routine for drawing single JPEG MCU
+            let draw_mcu = &mut |row_r: Rect, row_bitmap: BitmapView| {
+                // Draw a row of decoded MCUs
+                canvas.draw_bitmap(row_r.translate(bounds.top_left().into()), row_bitmap);
+                // Return true if we are not done yet
+                row_r.y1 < clip.y1
+            };
+
+            unwrap!(
+                cache
+                    .jpeg()
+                    .decompress_row(self.jpeg, self.scale, clip.y0, draw_mcu),
+                "Invalid JPEG"
+            );
+        } else {
+            // Draw JPEG with blurring effect
+            let jpeg_size = self.bounds(cache).size();
+
+            // Get a single line working bitmap
+            let buff = &mut unwrap!(cache.image_buff(), "No image buffer");
+            let mut slice = unwrap!(
+                Bitmap::new(
+                    BitmapFormat::RGB565,
+                    None,
+                    Offset::new(jpeg_size.x, 1),
+                    None,
+                    &buff[..]
+                ),
+                "Too small buffer"
+            );
+
+            // Get the blurring algorithm instance
+            let mut blur_cache = cache.blur();
+            let (blur, blur_tag) =
+                unwrap!(blur_cache.get(jpeg_size, self.blur_radius, self.blur_tag));
+            self.blur_tag = Some(blur_tag);
+
+            if let Some(y) = blur.push_ready() {
+                // A function for drawing a row of JPEG MCUs
+                let draw_row = &mut |row_r: Rect, jpeg_slice: BitmapView| {
+                    loop {
+                        if let Some(y) = blur.push_ready() {
+                            if y < row_r.y1 {
+                                // should never fail
+                                blur.push(unwrap!(jpeg_slice.row(y - row_r.y0)));
+                            } else {
+                                return true; // need more data
+                            }
+                        }
+
+                        if let Some(y) = blur.pop_ready() {
+                            blur.pop(unwrap!(slice.row_mut(0)), Some(self.dim)); // should never fail
```

Comment:
This is currently failing in `test_sd_protect_unlock` on T3T1.

Reply:
thanks for catching, the bug was introduced in f7562ac92409679a30c63e3ad9bc667f74f897e1 and fixed in c66be66eba97cd874b34d3b963caeca703de2536